source/base/data_out_base.cc

```diff
@@ -8191,68 +8192,122 @@ DataOutInterface<dim, spacedim>::create_xdmf_entry(
+  // The implementation is a bit complicated because we are supposed to return
+  // the correct data on rank 0 and an empty object on all other ranks but all
+  // information (for example the attributes) are only available on ranks that
+  // have any cells.
+  // We will identify the smallest rank that has data and then communicate
+  // from this rank to rank 0 (if they are different ranks).
+
+  const bool have_data = (data_filter.n_nodes() > 0);
+  MPI_Comm split_comm;
+  {
+    const int key = myrank;
+    const int color = (have_data ? 1 : 0);
+    const int ierr = MPI_Comm_split(comm, color, key, &split_comm);
+    AssertThrowMPI(ierr);
+  }
+
+  const bool am_i_first_rank_with_data =
+    have_data && (Utilities::MPI::this_mpi_process(split_comm) == 0);
+
+  ierr = MPI_Comm_free(&split_comm);
+  AssertThrowMPI(ierr);
+
+  const int tag = 47381;
+
+  // Output the XDMF file only on the root process of all ranks with data:
+  if (am_i_first_rank_with_data)
     {
       const auto &patches = get_patches();
       Assert(patches.size() > 0, DataOutBase::ExcNoPatches());
+
       // We currently don't support writing mixed meshes:
-#ifdef DEBUG
+#  ifdef DEBUG
       for (const auto &patch : patches)
         Assert(patch.reference_cell == patches[0].reference_cell,
                ExcNotImplemented());
-#endif
-
+#  endif
 
-      XDMFEntry    entry(h5_mesh_filename,
+      XDMFEntry entry(h5_mesh_filename,
                       h5_solution_filename,
                       cur_time,
                       global_node_cell_count[0],
                       global_node_cell_count[1],
                       dim,
                       spacedim,
                       patches[0].reference_cell);
-      unsigned int n_data_sets = data_filter.n_data_sets();
+      const unsigned int n_data_sets = data_filter.n_data_sets();
 
-      // The vector names generated here must match those generated in the HDF5
-      // file
-      unsigned int i;
-      for (i = 0; i < n_data_sets; ++i)
+      // The vector names generated here must match those generated in
+      // the HDF5 file
+      for (unsigned int i = 0; i < n_data_sets; ++i)
         {
           entry.add_attribute(data_filter.get_data_set_name(i),
                               data_filter.get_data_set_dim(i));
         }
 
+      if (myrank != 0)
+        {
+          // send to rank 0
+          const std::vector<char> buffer = Utilities::pack(entry, false);
+          ierr = MPI_Send(buffer.data(), buffer.size(), MPI_CHAR, 0, tag, comm);
+          AssertThrowMPI(ierr);
+
+          return {};
+        }
+
       return entry;
     }
-  else
+
+  if (myrank == 0 && !am_i_first_rank_with_data)
     {
-      return {};
+      // receive the XDMF data on rank 0 if we don't have it...
+
+      MPI_Status status;
+      int ierr = MPI_Probe(MPI_ANY_SOURCE, tag, comm, &status);
+      AssertThrowMPI(ierr);
+
+      int len;
+      ierr = MPI_Get_count(&status, MPI_BYTE, &len);
```

Comment:
Good question. I don't think it makes a practical difference as both have the same size and no type conversion happens here. Is zlib compressed data a collection of bytes or more like a string of chars?

If I was sending an std::string I would use MPI_CHAR, but here it is less clear to me. I am happy to change to MPI_CHAR if anybody else has thoughts about this.

Reply:
My machine's copy of zlib uses unsigned chars as their byte type so we should use `MPI_BYTE` but I doubt that it really matters in practice - we aren't ever going to work correctly on mixed architecture clusters since none of our serialization code knows how to handle it.